processing/src/main/java/io/druid/segment/StringDimensionHandler.java

```diff
@@ -0,0 +1,1248 @@
+      this.idToIndex = new int[length];
+      this.indexToId = new int[length];
+      int index = 0;
+      for (Integer id : sortedMap.values()) {
+        idToIndex[id] = index;
+        indexToId[index] = id;
+        index++;
+      }
+    }
+
+    public int size()
+    {
+      return sortedVals.size();
+    }
+
+    public int getUnsortedIdFromSortedId(int index)
+    {
+      return indexToId[index];
+    }
+
+    public int getSortedIdFromUnsortedId(int id)
+    {
+      return idToIndex[id];
+    }
+
+    public String getValueFromSortedId(int index)
+    {
+      return Strings.emptyToNull(sortedVals.get(index));
+    }
+  }
+
+  public class StringDimensionIndexer implements DimensionIndexer<Integer, String>
+  {
+    private DimensionDictionary dimLookup;
+    private SortedDimensionDictionary sortedLookup;
+    private Object lock;
+
+    public StringDimensionIndexer(Object lock)
+    {
+      this.lock = lock;
+      this.dimLookup = new DimensionDictionary(this.lock);
+    }
+
+    @Override
+    public void addNullLookup()
+    {
+      int oldDictSize = dimLookup.size();
+      dimLookup.add(null);
+      if (oldDictSize != dimLookup.size()) {
+        sortedLookup = dimLookup.sort();
+      }
+    }
+
+    @Override
+    public Comparable[] processRowValsToIndexKey(Object dimValues)
+    {
+      final Integer[] encodedDimensionValues;
+      int oldDictSize = dimLookup.size();
+      try {
+        if (dimValues == null) {
+          dimLookup.add(null);
+          encodedDimensionValues = null;
+        } else if (dimValues instanceof List) {
+          List<Object> dimValuesList = (List) dimValues;
+
+          // Sort multival row by their unencoded values first.
+          final String[] dimensionValues = new String[dimValuesList.size()];
+          for (int i = 0; i < dimValuesList.size(); i++) {
+            dimensionValues[i] = STRING_TRANSFORMER.apply(dimValuesList.get(i));
+          }
+          Arrays.sort(dimensionValues, UNENCODED_COMPARATOR);
+
+          encodedDimensionValues = new Integer[dimensionValues.length];
+          for (int i = 0; i < dimensionValues.length; i++) {
+            encodedDimensionValues[i] = dimLookup.add(dimensionValues[i]);
+          }
+        } else {
+          String transformedVal = STRING_TRANSFORMER.apply(dimValues);
+          encodedDimensionValues = new Integer[]{dimLookup.add(transformedVal)};
+        }
+      }
+      catch (ParseException pe) {
+        throw new ParseException(pe.getMessage() + dimensionName);
+      }
+
+      // If dictionary size has changed, the sorted lookup is no longer valid.
+      if (oldDictSize != dimLookup.size()) {
+        sortedLookup = null;
+      }
+
+      return encodedDimensionValues;
+    }
+
+    @Override
+    public String getActualValue(Integer intermediateValue, boolean idSorted)
+    {
+      if (idSorted) {
+        if (sortedLookup == null) {
+          sortedLookup = dimLookup.sort();
+        }
```

Comment:
this pattern of null check and then setting sortedLookup is repeated throughout the code, can you create a private method getSortedLookup() that encapsulates this pattern and returns the non-null sortedLookup ?